vm/ubpf_loader.c

```diff
@@ -220,13 +226,66 @@ ubpf_load_elf(struct ubpf_vm* vm, const void* elf, size_t elf_size, char** errms
                 goto error;
             }
 
-            unsigned int imm = ubpf_lookup_registered_function(vm, sym_name);
-            if (imm == -1) {
-                *errmsg = ubpf_error("function '%s' not found", sym_name);
+            switch (ELF64_R_TYPE(r.r_info)) {
+            // Perform map relocation.
+            case R_BPF_64_64: {
+                if (sym.st_shndx > ehdr->e_shnum) {
+                    *errmsg = ubpf_error("bad section index");
+                    goto error;
+                }
+                struct section* map = &sections[sym.st_shndx];
+                if (map->shdr->sh_type != SHT_PROGBITS || map->shdr->sh_flags != (SHF_ALLOC | SHF_WRITE)) {
+                    *errmsg = ubpf_error("bad map section");
+                    goto error;
+                }
+
+                if (sym.st_size + sym.st_value > map->size) {
+                    *errmsg = ubpf_error("bad map size");
+                    goto error;
+                }
+
+                struct ebpf_inst* inst = (struct ebpf_inst*)(text_copy + r.r_offset);
+                struct ebpf_inst* inst2 = (struct ebpf_inst*)(text_copy + r.r_offset + sizeof(struct ebpf_inst));
+                if (inst->opcode != EBPF_OP_LDDW) {
+                    *errmsg = ubpf_error("bad relocation instruction");
+                    goto error;
+                }
+                if (r.r_offset + sizeof(struct ebpf_inst) * 2 > text->size) {
+                    *errmsg = ubpf_error("bad relocation offset");
+                    goto error;
+                }
+
+                if (!vm->data_relocation_function) {
+                    *errmsg = ubpf_error("map relocation function not set");
+                    goto error;
+                }
+
+                const uint8_t* data = sym.st_value + map->data;
+                uint64_t size = sym.st_size;
+                uint64_t imm =
+                    vm->data_relocation_function(vm->data_relocation_user_data, data, size, sym_name, sym.st_value);
+                inst->imm = (uint32_t)imm;
```

Comment:
Just wanted to confirm: It looks like you are passing the size of the symbol here and not the size of the section. (see above).